platform/viewer/src/connectedComponents/ViewerRetrieveStudyData.js

```diff
@@ -1,77 +1,118 @@
-import React, { Component } from "react";
-import { metadata, studies, utils } from "@ohif/core";
+import React, { Component } from 'react';
+import { metadata, studies, utils, log } from '@ohif/core';
 
-import ConnectedViewer from "./ConnectedViewer.js";
-import PropTypes from "prop-types";
-import { extensionManager } from "./../App.js";
+import ConnectedViewer from './ConnectedViewer.js';
+import PropTypes from 'prop-types';
+import { extensionManager } from './../App.js';
 
-const { OHIFStudyMetadata } = metadata;
+const { OHIFStudyMetadata, OHIFSeriesMetadata } = metadata;
 const { retrieveStudiesMetadata } = studies;
 const { studyMetadataManager, updateMetaDataManager } = utils;
 
 class ViewerRetrieveStudyData extends Component {
   static propTypes = {
     studyInstanceUids: PropTypes.array.isRequired,
     seriesInstanceUids: PropTypes.array,
-    server: PropTypes.object
+    server: PropTypes.object,
   };
 
-  state = {
-    studies: null,
-    error: null
-  };
-
-  async componentDidMount() {
-    // TODO: Avoid using timepoints here
-    //const params = { studyInstanceUids, seriesInstanceUids, timepointId, timepointsFilter={} };
-    const { studyInstanceUids, seriesInstanceUids, server } = this.props;
-
-    try {
-      const studies = await retrieveStudiesMetadata(
-        server,
-        studyInstanceUids,
-        seriesInstanceUids
-      );
+  constructor(props) {
+    super(props);
+    this.isReady = false;
+    this.state = {
+      studies: null,
+      error: null,
+    };
+  }
 
-      // Render the viewer when the data is ready
-      // TODO: CLEAR THIS SOMEWHERE ELSE
-      studyMetadataManager.purge();
+  triggerDataLoad() {
+    const { props } = this;
+    retrieveStudiesMetadata(
+      props.server,
+      props.studyInstanceUids,
+      props.seriesInstanceUids
+    ).then(
+      studies => {
+        if (this.isReady) {
+          this.setStudies(studies);
+        }
+      },
+      error => {
+        if (this.isReady) {
+          this.setState({ error: true });
+        }
+        log.error(error);
+      }
+    );
+  }
 
+  setStudies(givenStudies) {
+    if (Array.isArray(givenStudies) && givenStudies.length > 0) {
+      const sopClassHandlerModules =
+        extensionManager.modules['sopClassHandlerModule'];
       // Map studies to new format, update metadata manager?
-      const updatedStudies = studies.map(study => {
+      const studies = givenStudies.map(study => {
         const studyMetadata = new OHIFStudyMetadata(
           study,
           study.studyInstanceUid
         );
-        const sopClassHandlerModules =
-          extensionManager.modules["sopClassHandlerModule"];
-
-        study.displaySets =
-          study.displaySets ||
-          studyMetadata.createDisplaySets(sopClassHandlerModules);
+        if (!study.displaySets) {
```

Comment:
Did you not like the previous syntax?

```js
study.displaySets =
          study.displaySets ||
          studyMetadata.createDisplaySets(sopClassHandlerModules);
```

Reply:
The semantic of:
`if (!a) { a = b }`
is different from:
`a = a || b`
as the former is a conditional assignment operation and the later is an unconditional one.
This changes though when we have a scenario like:
`if (b) { a = b } else { a = c }`
where the assignment to the variable "a" always takes place making it equivalent to:
`a = b || c`
In this specific situation it does not make much difference but if study.displaySets was a setter instead of a regular property this change would result in no additional calls that setter.
The following code snippet illustrates this:
`
var o = {
  count: 0,
  set a(v) {
    this._a = v;
    this.count++;
  },
  get a() {
    return this._a;
  }
};
if (!o.a) o.a = 'a';
if (!o.a) o.a = 'b';
if (!o.a) o.a = 'c';
if (!o.a) o.a = 'd';
console.log(o.a); // 'a'
console.log(o.count); // 1
o.a = o.a || 'a';
o.a = o.a || 'b';
o.a = o.a || 'c';
o.a = o.a || 'd';
console.log(o.a); // 'a'
console.log(o.count); // 5
`

Reply:
Ooh. That's interesting.
Is this an optimization JS makes under the hood?

Reply:
No... This cannot be optimized because as I said the semantic is different.
Changing from `if (!o.a) o.a = b` to `o.a = o.a || b` and vice-versa may have different side effects as the gist in the link above demonstrates.

Reply:
I follow -- I was just trying to understand how/why we see benefits with `if (!o.a) o.a = b`
I get that there is a benefit; it's just not immediately obvious to me _why_ it's only called once if repeatedly checked.